client/trino-client/src/main/java/io/trino/client/MultiThreadedQuery.java

```diff
@@ -0,0 +1,267 @@
+    Semaphore allResultsReady;
+    LocalDateTime start;
+    String nextUri;
+    Optional<String> user;
+
+    public MultiThreadedQuery(ClientSession session, Headers headers, StatementClient caller, Optional<String> user)
+    {
+        this.start = LocalDateTime.now();
+        this.client = caller;
+        this.session = session;
+        this.user = user;
+        this.nextUri = headers.get(HEADER_NEXTURI);
+//        System.out.println("DEBUG: " + nextUri);
+        this.allResultsReady = new Semaphore(1);
+        try {
+            this.allResultsReady.acquire();
+        }
+        catch (Exception e) {
+            e.printStackTrace();
+        }
+
+        Results newResults = new Results(0);
+        newResults.nextUriReady(nextUri);
+        this.results = newResults;
+        this.currentResults = newResults;
+
+        Thread thread = new Thread(this);
+        thread.start();
+    }
+
+    /**
+     * Starts creating a Results linked list
+     */
+    @Override
+    public void run()
+    {
+        ExecutorService executorService = Executors.newFixedThreadPool(8);
+        ConnectionPool pool = new ConnectionPool(10, 5, TimeUnit.SECONDS);
+
+        String nextUri = this.nextUri;
+        Results newResults = this.results;
+        int counter = 1;
+
+        while (nextUri != null) {
+            try {
+                counter++;
+
+                String urlToDownload = nextUri;
+                Results res = newResults;
+                LocalDateTime start = LocalDateTime.now();
+                executorService.submit(() -> downloadData(urlToDownload, res));
+                res.waitForNextUri();
+                nextUri = res.nextUri;
+                if (nextUri != null) {
+                    newResults = new Results(counter);
+                    newResults.nextUri = nextUri;
+                    results.next = newResults;
+                    results = newResults;
+                }
+            }
+            catch (Exception e) {
+                e.printStackTrace();
+                nextUri = null;
+            }
+        }
+
+        LocalDateTime end = LocalDateTime.now();
+        Duration duration = Duration.between(start, end);
+        System.out.printf("NextUri trail completed in %2d.%06ds (%d result sets)\n", duration.getSeconds(), duration.getNano() / 1000, counter);
+
+        executorService.shutdown();
+        end = LocalDateTime.now();
+        duration = Duration.between(start, end);
+        System.out.printf("Fetch data completed in %2d.%06ds\n", duration.getSeconds(), duration.getNano() / 1000);
+    }
+
+    /**
+     * Downloads a given URI, parse it and store the results
+     * @param nextUri the URI to download
+     * @param results the Results object where to store the results
+     */
+    private void downloadData(String nextUri, Results results)
+    {
+        try {
+            HttpUrl url = HttpUrl.get(nextUri);
+            OkHttpClient client = new OkHttpClient.Builder()
```

Comment:
Don't create new httpClient object. This will not work for secure connections.
Instead, use httpClient object from StatementClientV1. More details on how to use httpClient is listed here - https://medium.com/@leandromazzuquini/if-you-are-using-okhttp-you-should-know-this-61d68e065a2b

Reply:
The reason a new `httpClient` is created is because it's adding a network interceptor (line 136). Is it possible to temporarily add it to an existing object?

Reply:
Yes.

How about this:
`            OkHttpClient client = httpClient.newBuilder()
                    .addNetworkInterceptor(new HeadersInterceptor(results))
                    .build();
`

httpClient is passed from StatementClientV1.

Reply:
Ah, that works. Fixed.